File: cpp/test/behaviour/steps/common/concept/attribute_steps.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <cstdlib>

#include "common.hpp"
#include "concept_utils.hpp"
#include "steps.hpp"
#include "utils.hpp"

namespace TypeDB::BDD {

using namespace cucumber::messages;

cucumber_bdd::StepCollection<Context> attributeSteps = {
    BDD_STEP("attribute\\((\\S+)\\) get instances contain: \\$(\\S+)", {
        ASSERT_TRUE(containsInstance(collect(attrType(context, matches[1].str())->getInstances(context.transaction())), context.vars[matches[2].str()].get()));
    }),
    BDD_STEP("attribute \\$(\\S+) has value type: (\\w+)", {
        ASSERT_EQ(parseValueType(matches[2].str()), context.vars[matches[1].str()]->asAttribute()->getType()->getValueType());
    }),

    // Create instance
    BDD_STEP("attribute\\((\\S+)\\) as\\(boolean\\) put: (true|false); throws exception", {
        DRIVER_THROWS("", {
            attrType(context, matches[1].str())->put(context.transaction(), parseBoolean(matches[2].str())).get();
        });
    }),
    BDD_STEP("attribute\\((\\S+)\\) as\\(long\\) put: (\\d+); throws exception", {
        DRIVER_THROWS("", {
            attrType(context, matches[1].str())->put(context.transaction(), parseBoolean(matches[2].str())).get();
        });
    }),
    BDD_STEP("attribute\\((\\S+)\\) as\\(double\\) put: ([0-9\\.]+); throws exception", {
        DRIVER_THROWS("", {
            attrType(context, matches[1].str())->put(context.transaction(), parseBoolean(matches[2].str())).get();
        });
    }),
    BDD_STEP("attribute\\((\\S+)\\) as\\(string\\) put: (.*); throws exception", {
        DRIVER_THROWS("", {
            attrType(context, matches[1].str())->put(context.transaction(), parseBoolean(matches[2].str())).get();
        });
    }),
    BDD_STEP("attribute\\((\\S+)\\) as\\(datetime\\) put: ([0-9TZ:\\- ]+); throws exception", {
        DRIVER_THROWS("", {
            attrType(context, matches[1].str())->put(context.transaction(), parseBoolean(matches[2].str())).get();
        });
    }),

    BDD_STEP("\\$(.+) = attribute\\((.+)\\) as\\(boolean\\) put: (true|false)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->put(context.transaction(), parseBoolean(matches[3].str())).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(long\\) put: (\\d+)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->put(context.transaction(), parseLong(matches[3].str())).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(double\\) put: ([0-9\\.]+)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->put(context.transaction(), parseDouble(matches[3].str())).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(string\\) put: (.*)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->put(context.transaction(), matches[3].str()).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(datetime\\) put: ([0-9TZ:\\- ]+)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->put(context.transaction(), parseDateTime(matches[3].str())).get();
    }),

    // Read instance
    BDD_STEP("attribute \\$(\\w+) has boolean value: (true|false)", {
        ASSERT_EQ(parseBoolean(matches[2].str()), context.vars[matches[1].str()]->asAttribute()->getValue()->asBoolean());
    }),
    BDD_STEP("attribute \\$(\\w+) has long value: (\\d+)", {
        ASSERT_EQ(parseLong(matches[2].str()), context.vars[matches[1].str()]->asAttribute()->getValue()->asLong());
    }),
    BDD_STEP("attribute \\$(\\w+) has double value: ([0-9\\.]+)", {
        ASSERT_EQ(parseDouble(matches[2].str()), context.vars[matches[1].str()]->asAttribute()->getValue()->asDouble());
    }),
    BDD_STEP("attribute \\$(\\w+) has string value: (.*)", {
        ASSERT_EQ(matches[2].str(), context.vars[matches[1].str()]->asAttribute()->getValue()->asString());
    }),
    BDD_STEP("attribute \\$(\\w+) has datetime value: ([0-9TZ:\\- ]+)", {
        ASSERT_EQ(parseDateTime(matches[2].str()), context.vars[matches[1].str()]->asAttribute()->getValue()->asDateTime());
    }),

    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(boolean\\) get: (true|false)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->get(context.transaction(), parseBoolean(matches[3].str())).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(long\\) get: (\\d+)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->get(context.transaction(), parseLong(matches[3].str())).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(double\\) get: ([0-9\\.]+)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->get(context.transaction(), parseDouble(matches[3].str())).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(string\\) get: (.*)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->get(context.transaction(), matches[3].str()).get();
    }),
    BDD_STEP("\\$(\\w+) = attribute\\((\\S+)\\) as\\(datetime\\) get: ([0-9TZ:\\- ]+)", {
        context.vars[matches[1].str()] = attrType(context, matches[2].str())->get(context.transaction(), parseDateTime(matches[3].str())).get();
    }),

    // Owners
    BDD_STEP("attribute \\$(\\S+) get owners contain: \\$(\\S+)", {
        auto owners = collect(context.vars[matches[1].str()]->asAttribute()->getOwners(context.transaction()));
        ASSERT_TRUE(containsInstance(owners, context.vars[matches[2].str()].get()));
    }),
    BDD_STEP("attribute \\$(\\S+) get owners do not contain: \\$(\\S+)", {
        auto owners = collect(context.vars[matches[1].str()]->asAttribute()->getOwners(context.transaction()));
        ASSERT_FALSE(containsInstance(owners, context.vars[matches[2].str()].get()));
    }),
};

}  // namespace TypeDB::BDD

File: nodejs/test/behaviour/concept/thing/attribute/AttributeSteps.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {When} from "@cucumber/cucumber";
import {Concept} from "../../../../../dist";
import {tx} from "../../../connection/ConnectionStepsBase";
import {assertThrows} from "../../../util/Util";
import {get, put} from "../ThingSteps";
import assert = require("assert");
import ValueType = Concept.ValueType;

When("attribute\\({type_label}) get instances contain: {var}", async (typeLabel: string, var0: string) => {
    assert(await (await tx().concepts.getAttributeType(typeLabel)).getInstances(tx()).some(x => x.equals(get(var0))));
});

When("attribute {var} get owners contain: {var}", async (var1: string, var2: string) => {
    assert(await get(var1).asAttribute().getOwners(tx()).some(x => x.equals(get(var2))));
});

When("attribute {var} get owners do not contain: {var}", async (var1: string, var2: string) => {
    assert(!(await get(var1).asAttribute().getOwners(tx()).some(x => x.equals(get(var2)))));
});

When("attribute {var} has value type: {value_type}", async (var0: string, valueType: ValueType) => {
    assert.strictEqual(valueType, get(var0).asAttribute().type.valueType);
});

When("{var} = attribute\\({type_label}) as\\(boolean) put: {bool}", async (var0: string, typeLabel: string, value: boolean) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).putBoolean(tx(), value));
});

When("attribute\\({type_label}) as\\(boolean) put: {bool}; throws exception", async (typeLabel: string, value: boolean) => {
    await assertThrows(async () => await (await tx().concepts.getAttributeType(typeLabel)).putBoolean(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(long) put: {int}", async (var0: string, typeLabel: string, value: number) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).putLong(tx(), value));
});

When("attribute\\({type_label}) as\\(long) put: {int}; throws exception", async (typeLabel: string, value: number) => {
    await assertThrows(async () => await (await tx().concepts.getAttributeType(typeLabel)).putLong(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(double) put: {float}", async (var0: string, typeLabel: string, value: number) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).putDouble(tx(), value));
});

When("attribute\\({type_label}) as\\(double) put: {float}; throws exception", async (typeLabel: string, value: number) => {
    await assertThrows(async () => await (await tx().concepts.getAttributeType(typeLabel)).putDouble(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(string) put: {word}", async (var0: string, typeLabel: string, value: string) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).putString(tx(), value));
});

When("attribute\\({type_label}) as\\(string) put: {word}; throws exception", async (typeLabel: string, value: string) => {
    await assertThrows(async () => await (await tx().concepts.getAttributeType(typeLabel)).putString(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(datetime) put: {datetime}", async (var0: string, typeLabel: string, value: Date) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).putDateTime(tx(), value));
});

When("attribute\\({type_label}) as\\(datetime) put: {datetime}; throws exception", async (typeLabel: string, value: Date) => {
    await assertThrows(async () => await (await tx().concepts.getAttributeType(typeLabel)).putDateTime(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(boolean) get: {bool}", async (var0: string, typeLabel: string, value: boolean) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).getBoolean(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(long) get: {int}", async (var0: string, typeLabel: string, value: number) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).getLong(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(double) get: {float}", async (var0: string, typeLabel: string, value: number) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).getDouble(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(string) get: {word}", async (var0: string, typeLabel: string, value: string) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).getString(tx(), value));
});

When("{var} = attribute\\({type_label}) as\\(datetime) get: {datetime}", async (var0: string, typeLabel: string, value: Date) => {
    put(var0, await (await tx().concepts.getAttributeType(typeLabel)).getDateTime(tx(), value));
});

When("attribute {var} has boolean value: {bool}", async (var0: string, value: boolean) => {
    assert.strictEqual(value, get(var0).asAttribute().value);
});

When("attribute {var} has long value: {int}", async (var0: string, value: number) => {
    assert.strictEqual(value, get(var0).asAttribute().value);
});

When("attribute {var} has double value: {float}", async (var0: string, value: number) => {
    assert.strictEqual(value, get(var0).asAttribute().value);
});

When("attribute {var} has string value: {word}", async (var0: string, value: string) => {
    assert.strictEqual(value, get(var0).asAttribute().value);
});

When("attribute {var} has datetime value: {datetime}", async (var0: string, value: Date) => {
    assert.strictEqual(value.getTime(), (get(var0).asAttribute().value as Date).getTime());
});
